File: SinglyLinkedList.js
class SinglyLinkedListNode {
  constructor(data) {
    this.data = data;
    this.next = null;
  }
}

class SinglyLinkedList {
  constructor() {
    this.head = null;
  }

  isEmpty() {
    // or return this.head === null
    return this.head === null;
  }

  // insert at the head of the SLL
  // Time complexity: O(1)
  insert(value) {
    if (this.isEmpty()) {
      this.head = new SinglyLinkedListNode(value);
    } else {
      const oldHead = this.head;
      const newHead = new SinglyLinkedListNode(value);
      newHead.next = oldHead;
      this.head = newHead;
    }
  }

  // delete node whose data = given value
  // Time complexity: O(n)
  remove(value) {
    let currentHead = this.head;
    // if the removed node is the head
    if (this.head.data === value) {
      this.head = currentHead.next;
    } else {
      // in the middle of the SLL
      let prev = currentHead;
      while (currentHead) {
        if (currentHead.data === value) {
          prev.next = currentHead.next;
          break;
        } else {
          prev = currentHead;
        }
        currentHead = currentHead.next;
      }
    }
  }

  // Time complexity: O(1)
  deleteAtHead() {
    let toReturn = null;
    if (this.head !== null) {
      toReturn = this.head.data;
      this.head = this.head.next;
    }
    return toReturn;
  }

  // Time complexity: O(n)
  find(value) {
    let currentHead = this.head;
    while (currentHead) {
      if (currentHead.data === value) return true;
      currentHead = currentHead.next;
    }
    return false;
  }

  traverse() {
    let path = "";
    for (let node = this.head; node !== null; node = node.next) {
      path += node.data + " -> ";
    }
    path += "null";
    return path;
  }
}
module.exports = SinglyLinkedList;

File: SinglyLinkedList.py
from typing import Optional

class ListNode:
  def __init__(self, val: int) -> None:
    self.val = val
    self.next: Optional[ListNode] = None
  
class SinglyLinkedList:
  def __init__(self) -> None:
    self.head: Optional[ListNode] = None

  def insert_head(self, val: int):
    if not self.head: 
      self.head = ListNode(val)
      return
    old_head = self.head
    self.head = ListNode(val)
    self.head.next = old_head
  
  def delete_by_value(self, val: int):
    if not self.head: return
    if self.head.val == val:
      self.head = self.head.next
    cur = self.head
    prev: Optional[ListNode] = None
    while cur:
      if cur.val == val:
        prev.next = cur.next 
      prev = cur
      cur = cur.next

  def delete_head(self) -> int or None:
    if not self.head: return
    old_head = self.head
    self.head = self.head.next
    return old_head.val
  
  def contains(self, val: int) -> bool:
    cur = self.head
    while cur:
      if cur.val == val: return True
      cur = cur.next
    return False

  def print_sll(self) -> str:
    cur = self.head
    path = ''
    while cur:
      path += str(cur.val) + ' -> '
      cur = cur.next
    path += 'None'
    return path
# Time complexity
# n: # of nodes in a SLL
# Operation               Analysis
# Insertion at head       O(1)
# Deletion at head        O(1)
# Deletion by a value     O(n)
# Find                    O(n)